Refactor Livisi Switch and Climate to inherit from a common base class

planbnet opened the pull request:
## Breaking change

## Proposed change

As discussed in #88886 create a common base class for all livisi entities and refactor switch and climate to use it.
By using the base class where `_attr_has_entity_name = True`, the switch entity now also gets its name from the device, as proposed in #88891. 

This means that all livisi devices will have the same name as the entity, so we only set the name on the device and don't set _attr_name on the entity. There's just one exception: Climate device are automatically generated for each room (so they will always have a room set, the None check in the code is just defensive coding) - so for those `use_room_as_device_name` it true to give the device the name of the room. The entity name for those will always be some localized value of "Climate". So currently:

* a climate device named XYZ in a room ABC will have entity name XYZ and device name ABC 
* a climate device named XYZ without a room will have entity name None and device name XYZ (although this won't happen, because climate devices always have a room)
* a switch device named XYZ in a room ABC will have entity name None and device name XYZ 
* a switch device named XYZ without a room will have entity name None and device name XYZ 
* other devices will probably be handled the same way as switch in the future, becasue as far as I know Climate devices are the only virtual devices that are generated per-room

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

planbnet committed:
Refactor Livisi entities to inherit from a common base class

planbnet committed:
Add livisi_entity to .coveragerc

epenet commented on homeassistant/components/livisi/climate.py:
```diff
@@ -97,39 +67,25 @@ class LivisiClimate(CoordinatorEntity[LivisiDataUpdateCoordinator], ClimateEntit
+        # the livisi setup, as each livisi room gets exactly one VRCC device. The entity
+        # name will always be some localized value of "Climate", so the full element name
+        # in homeassistent will be in the form of "Bedroom Climate"
+        self._attr_device_info["name"] = self._attr_device_info["suggested_area"]
```
This is confusing.
I think you should instead add an argument (eg. `*, use_room_as_device_name: bool = False`) to the entity `__init__` method and move the comment there.

Then here you can simply use:
`super().__init__(config_entry, coordinator, device, use_room_as_device_name=True)`

planbnet replied:
Great suggestion. Done.

planbnet committed:
Device location can be None

planbnet committed:
Add use_room_as_device_name argument to constructor of LivisiEntity

When initializing, set entity name attribute only if device name differs (i.e. use_room_as_device_name=True).

planbnet committed:
re-add comment for special handling of climate device names

epenet commented on homeassistant/components/livisi/livisi_entity.py:
```diff
@@ -28,6 +28,7 @@ def __init__(
         config_entry: ConfigEntry,
         coordinator: LivisiDataUpdateCoordinator,
         device: dict[str, Any],
+        use_room_as_device_name=False,
```
I suggest to make this a keyword only argument, which will give more flexibility:
```suggestion
        *,
        use_room_as_device_name=False,
```

planbnet committed:
Add explicit type to constructur argument

epenet reviewed (changes requested):
Could you please extra notes in the PR description regarding the naming scheme?
For example:
- a climate device named XYZ in a room ABC will have entity name ___ and device name ___
- a climate device named XYZ without a room will have entity name ___ and device name ___
- a switch device named XYZ in a room ABC will have entity name ___ and device name ___
- a switch device named XYZ without a room will have entity name ___ and device name ___

It's kind of hard to work out what is happening right now.

planbnet committed:
Make use_room_as_device_name a keyword only arg

epenet commented on .coveragerc:
```diff
@@ -644,6 +644,7 @@ omit =
     homeassistant/components/livisi/climate.py
     homeassistant/components/livisi/coordinator.py
     homeassistant/components/livisi/switch.py
+    homeassistant/components/livisi/livisi_entity.py
```
I also suggest you rename this file to `entity.py`. We already know we're in `livisi` domain...

planbnet committed:
rename livisi_entity.py to entity.py

planbnet committed:
change livisi_entity.py to entity.py in coveragerc

planbnet commented:
All changes done as proposed. Btw, you have a great contribution workflow and I like your focus on code clarity and quality. As this is the first time for me contributing in a large python codebase, I get to learn quite a lot about here... Thanks!

epenet commented:
> a climate device named XYZ without a room will have entity name XYZ and device name XYZ (although this won't happen, because climate devices always have a room)

When I read the code, I think entity name would be `None` in this edge case. I think the PR description is wrong here.

planbnet commented:
> When I read the code, I think entity name would be `None` in this edge case. I think the PR description is wrong here.

You're right. I corrected that.

epenet commented on homeassistant/components/livisi/entity.py:
```diff
@@ -0,0 +1,88 @@
+        use_room_as_device_name: bool = False,
+    ) -> None:
+        """Initialize the common properties of a Livisi device."""
+        self.config_details: Mapping[str, Any] = device["config"]
```
This is only used within init method, and only to set the name.
You could simply use `name = device["config"]["name"]`
Or if you think it is useful for the type hint you can just remove  `self.` assignment

epenet commented on homeassistant/components/livisi/entity.py:
```diff
@@ -0,0 +1,88 @@
+        self.config_details: Mapping[str, Any] = device["config"]
+
+        self.aio_livisi = coordinator.aiolivisi
+        self.config_entry = config_entry
```
There is no need to hold a reference to `config_entry`. It is only used here inside `__init__`.
```suggestion
```

epenet commented on homeassistant/components/livisi/entity.py:
```diff
@@ -0,0 +1,88 @@
+
+        name = self.config_details["name"]
+        unique_id = device["id"]
+        manufacturer = device["manufacturer"]
+        device_type = device["type"]
```
I don't think you need to create variables here. Just set them directly in the device_info.
```python
        manufacturer=device["manufacturer"],
        model=device["type"],
```

epenet reviewed (changes requested):
Last few comments and then I think it's good to go.

planbnet committed:
Code quality improvements as suggested in PR

epenet reviewed (approved):
LGTM 👍 
Thanks @planbnet

epenet commented:
(once CI passes - see `isort` issue)

epenet commented on .coveragerc:
```diff
@@ -644,6 +644,7 @@ omit =
     homeassistant/components/livisi/climate.py
     homeassistant/components/livisi/coordinator.py
     homeassistant/components/livisi/switch.py
+    homeassistant/components/livisi/entity.py
```
Just spotted this: can it be kept in alphabetical order

planbnet committed:
sort .coveragerc

epenet commented on homeassistant/components/livisi/switch.py:
```diff
@@ -8,18 +8,16 @@
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
-from homeassistant.helpers.update_coordinator import CoordinatorEntity
 
 from .const import (
```
Did you miss this `isort` issue?
`from .const import DOMAIN, LIVISI_STATE_CHANGE, LOGGER, PSS_DEVICE_TYPE`

planbnet replied:
Sorry, my fault. I din't really understand what the system wanted to tell me :)

planbnet committed:
fix isort issue

planbnet committed:
fix all isort issues

epenet reviewed (approved):
LGTM 👍

MartinHjelmare commented on homeassistant/components/livisi/entity.py:
```diff
@@ -0,0 +1,81 @@
+from .coordinator import LivisiDataUpdateCoordinator
+
+
+class LivisiEntity(CoordinatorEntity[LivisiDataUpdateCoordinator], Entity):
```
We don't need to inherit `Entity`. The `CoordinatorEntity` already does that.

MartinHjelmare reviewed (commented):
Please address the comment in a new PR. Thanks!
